### PR TITLE
Implement threshold decision policy

### LINKED ISSUE #45: Implement ThresholdDecisionPolicy
needed-by #31

### PR DESCRIPTION
**Pull request by alpe**
Resolves #45

After #54

### CODE REVIEW

File: incubator/group/types.go [hunk 1]
```diff
@@ -45,11 +45,13 @@ type DecisionPolicyResult struct {
 	Final bool
 }
 
+// DecisionPolicy persistent ruleset to determine the result of election on a proposal.
```
Comment (aaronc) [inline incubator/group/types.go]:
```suggestion
// DecisionPolicy is persistent set of rules to determine the result of election on a proposal.
```

File: incubator/group/types.go [hunk 2]
```diff
@@ -45,11 +45,13 @@ type DecisionPolicyResult struct {
...
 type DecisionPolicy interface {
 	orm.Persistent
 	Allow(tally Tally, totalPower sdk.Dec, votingDuration time.Duration) (DecisionPolicyResult, error)
 }
 
+// Allow allow a proposal to pass when the threshold is exceeded by yes votes before the timeout.
```
Comment (aaronc) [inline incubator/group/types.go]:
```suggestion
// Allow allows a proposal to pass when the tally of yes votes equals or exceeds the threshold before the timeout.
```

File: incubator/group/types.proto [hunk 3]
```diff
@@ -113,7 +113,6 @@ message StdDecisionPolicy {
         ThresholdDecisionPolicy threshold = 1;
     }
 }
-// todo: we need to ensure that ThresholdDecisionPolicy can not be modified during a running proposal
 message ThresholdDecisionPolicy {
     // threshold is a fix weight value that must be exceeded for a proposal to succeed.
```
Comment (aaronc) [inline incubator/group/types.proto]:
```suggestion
    // threshold is a fix weight value that must be met or exceeded for a proposal to succeed.
```

File: incubator/group/types_test.go [hunk 4]
```diff
@@ -2,13 +2,154 @@ package group
...
+			srcPolicy: ThresholdDecisionPolicy{
+				Threshold: sdk.OneDec(),
+				Timout:    proto.Duration{Seconds: 1},
+			},
+			srcTally:          Tally{YesCount: sdk.NewDec(2)},
+			srcTotalPower:     sdk.NewDec(3),
+			srcVotingDuration: time.Second,
+			expResult:         DecisionPolicyResult{Allow: true, Final: true},
+		},
+		"reject when yes count equal to threshold": {
```
Comment (aaronc) [inline incubator/group/types_test.go]:
```suggestion
		"accept when yes count equal to threshold": {
```

File: incubator/group/types_test.go [hunk 5]
```diff
@@ -2,13 +2,154 @@ package group
...
+			srcPolicy: ThresholdDecisionPolicy{
+				Threshold: sdk.OneDec(),
+				Timout:    proto.Duration{Seconds: 1},
+			},
+			srcTally:          Tally{YesCount: sdk.OneDec(), NoCount: sdk.ZeroDec(), AbstainCount: sdk.ZeroDec(), VetoCount: sdk.ZeroDec()},
+			srcTotalPower:     sdk.NewDec(3),
+			srcVotingDuration: time.Second,
+			expResult:         DecisionPolicyResult{Allow: false, Final: false},
```
Comment (aaronc) [inline incubator/group/types_test.go]:
```suggestion
			expResult:         DecisionPolicyResult{Allow: true, Final: true},
```

File: incubator/group/types_test.go [hunk 6]
```diff
@@ -2,13 +2,154 @@ package group
...
+			srcPolicy: ThresholdDecisionPolicy{
+				Threshold: sdk.OneDec(),
+				Timout:    proto.Duration{Seconds: 1},
+			},
+			srcTally:          Tally{YesCount: sdk.ZeroDec(), NoCount: sdk.ZeroDec(), AbstainCount: sdk.ZeroDec(), VetoCount: sdk.ZeroDec()},
+			srcTotalPower:     sdk.NewDec(3),
+			srcVotingDuration: time.Second,
+			expResult:         DecisionPolicyResult{Allow: false, Final: false},
+		},
+		"reject as final when remaining votes can't cross threshold": {
```
Comment (aaronc) [inline incubator/group/types_test.go:52]: Update according to new definition of threshold

File: incubator/group/types_test.go [hunk 7]
```diff
@@ -2,13 +2,154 @@ package group
...
+			srcPolicy: ThresholdDecisionPolicy{
+				Threshold: sdk.OneDec(),
+				Timout:    proto.Duration{Seconds: 1},
+			},
+			srcTally:          Tally{YesCount: sdk.NewDec(2)},
+			srcTotalPower:     sdk.NewDec(3),
+			srcVotingDuration: time.Second + time.Nanosecond,
+			expResult:         DecisionPolicyResult{Allow: false, Final: true},
+		},
+		"abstain has not impact": {
```
Comment (aaronc) [inline incubator/group/types_test.go]:
```suggestion
		"abstain has no impact": {
```

**Review by aaronc — CHANGES_REQUESTED** [review]
The tests look good but I realize that the threshold policy is actually incorrectly specified. The threshold must be _met or_ exceeded. Just exceeding it is counter intuitive because in order to have a 1 out of 3 policy, you would need to set the threshold to 0. Looks like I messed up with my initial specification. My apologies!

**Comment by alpe** [comment]
I have modified the policy. With equal numbers we need to prohibit threshold==0, now so that we have a least 1 yes vote.

File: incubator/group/types.go [hunk 8]
```diff
@@ -45,13 +45,13 @@ type DecisionPolicyResult struct {
 	Final bool
 }
 
-// DecisionPolicy persistent ruleset to determine the result of election on a proposal.
+// DecisionPolicy is persistent set of rules to determine the result of election on a proposal.
```
Comment (aaronc) [inline incubator/group/types.go]:
```suggestion
// DecisionPolicy is the persistent set of rules to determine the result of election on a proposal.
```

**Review by aaronc — CHANGES_REQUESTED** [review]
Just found another minor doc type. Otherwise LGTM and ready to merge.

**Comment by alpe** [comment]
I had applied the change. Waiting for your approval (on github)

**Review by aaronc — APPROVED** [review]

**Pull request merged**